special soundness protocol for lookup (part 1)

chaosma opened the pull request:
* Implement SPS protocol for lookup (part 1)
* Multiple rounds of witness and challenges generation
* Integrate lookup arguments into (Relaxed)PlonkInstance/Witness

This part 1 only covers and be tested for the case with single custom gate and no lookup support. Will address multiple custom gates and lookup arguments in NIFS in part 2.

Impact on IVC Circuit by part 1: In (Relaxed)PlonkInstance, witness commits becomes a vector of commitments (renamed to W_commitments) instead of a single element (W_commitment)

chaosma commented on src/nifs.rs:
```diff
@@ -181,12 +175,13 @@ impl<C: CurveAffine, RO: ROTrait<C>> NIFS<C, RO> {
-        let mut U2 = U2;
         U2.absorb_into(ro);
-        if S.num_challenges > 1 {
+        /*
+        if S.num_challenges > 0 {
```
Depending on whether we have lookup/vector lookup etc, we might have different number of challenges to deal with. I leave it as a comment so I can pick it up later.

cyphersnake replied:
Add context for this commented lines

chaosma replied:
It has been removed in my next pr (working in progress)

chaosma commented on src/plonk/mod.rs:
```diff
@@ -374,39 +417,44 @@ impl<C: CurveAffine> RelaxedPlonkInstance<C> {
-            W,
-            E,
+        for sz in round_sizes.iter() {
+            let tmp = vec![F::ZERO; *sz];
+            W.push(tmp);
```
Witness: Vec<Vec<F>> for multi-round sps. Each element (vector) corresponds to a round in prover

cyphersnake replied:
I mean `tmp` not needed, just `W.push(vec!...`

chaosma commented on src/plonk/mod.rs:
```diff
@@ -449,11 +495,41 @@ impl<F: PrimeField> TableData<F> {
+    }
+
+    pub fn lookup_exprs(&self, cs: &ConstraintSystem<F>) -> Vec<Expression<F>> {
+        if self.lookup_arguments.is_none() {
+            vec![]
+        } else {
+            let lookup_arguments = self.lookup_arguments.clone().unwrap();
+            let mut combined = lookup_arguments.lookup_polys.clone();
+            combined.extend(lookup_arguments.log_derivative_lhs(cs));
+            combined
         }
```
changed to `if let Some(arg) = ...` pattern.

chaosma commented on src/plonk/mod.rs:
```diff
@@ -100,8 +111,7 @@ pub struct RelaxedPlonkInstance<C: CurveAffine> {
 pub struct RelaxedPlonkWitness<F: PrimeField> {
-    pub(crate) num_advice_columns: usize,
-    pub(crate) W: Vec<F>,
+    pub(crate) W: Vec<Vec<F>>,
```
docs added

chaosma commented on src/plonk/lookup.rs:
```diff
@@ -65,244 +52,158 @@ use crate::{
+        &self,
+        cs: &ConstraintSystem<F>,
+        lookup_index: usize,
+        cha_index: usize,
```
it is short for "challenge_index"

cyphersnake replied:
I wouldn't say it's a good abbreviation, as I didn't count it and an explanation needs to be added here. I would leave the full name to avoid confusion. 

What do you think?

chaosma replied:
sure

cyphersnake commented:
Part of #20

chaosma commented on src/plonk/lookup.rs:
```diff
@@ -65,244 +52,158 @@ use crate::{
+    }
+
+    /// evaluate the lookup and table expressions to get vector l and t
+    pub fn evaluate_ls_ts(&self, table: &TableData<F>, r: F) -> (Vec<Vec<F>>, Vec<Vec<F>>) {
```
added the definition of l and t. They can be calculated in one round. But we can also make it separate functions. Let me separate them.

chaosma commented on src/plonk/lookup.rs:
```diff
@@ -65,244 +52,158 @@ use crate::{
-        &self,
-        ss: &Structure<C>,
-    ) -> Vec<F> {
+    pub fn evaluate_m(&self, l: &[F], t: &[F]) -> Vec<F> {
```
rustdoc added

chaosma commented on src/plonk/lookup.rs:
since it is defined purely in some math formula, I don't think of a good name. I added the precise definition in front of the function definition, which is easy to lookup. But if you have some good suggestions, just let me know.

chaosma commented on src/plonk/mod.rs:
```diff
@@ -49,31 +48,44 @@ pub struct PlonkStructure<C: CurveAffine> {
     /// inst = [X0, X1]
     pub(crate) instance: Vec<C::ScalarExt>,
-    /// contains challenges i.e. random number generated by verifier or Fiat-Shamir hash
+    /// challenge is random number generated Fiat-Shamir hash
+    /// more specifically:
+    /// r1 (optional): combine vector lookup
+    /// r2 (optional): challenge to calculate h and g in log-derivative relation
+    /// r3 (optional): combine all relations
+    /// unlike protostar paper, we may add one extra round (r3) to combine all relations
     pub(crate) challenges: Vec<C::ScalarExt>,
```
refined the comments

chaosma commented:
@cyphersnake I changed W_commitment to W_commitments in IVC module and fixed compile and test errors accordingly. Some places (especially in tests) are quick hack, e.g. W_commitment changed to W_commitments[0], where I didn't dig into the detail. You can based on this to modify it in future PR if needed.

cyphersnake commented on src/ivc/fold_relaxed_plonk_instance_chip.rs:
```diff
@@ -586,7 +594,15 @@ where
         ro_circuit.absorb_base(WrapValue::Assigned(assigned_public_params_hash.clone()));
 
-        let assigned_W = assign_and_absorb_point!(self.W)?;
+        let assigned_W: Result<Vec<AssignedPoint<C>>, Error> = self
+            .W
+            .iter()
+            .map(|W| -> Result<AssignedPoint<C>, Error> {
+                let res = assign_and_absorb_point!(W)?;
+                Ok(res)
+            })
+            .collect();
+        let assigned_W = assigned_W?;
```
var not needed here
```rust
.collect()?;
```

chaosma replied:
I tried this before but got error, so I use above work around.

Error: `Try` is not implemented for `Vec<Result<ecc::AssignedPoint<C>,Error>`
Any idea how to fix it?

cyphersnake commented on src/ivc/fold_relaxed_plonk_instance_chip.rs:
```diff
@@ -659,7 +675,7 @@ where
 
         let assigned_instance_W_commitment_coordinates =
-            assign_and_absorb_point!(instance.W_commitment)?;
+            assign_and_absorb_point!(instance.W_commitments[0])?;
```
Why only zero?

cyphersnake replied:
```rust
let assigned_instance_W_commitment_coordinates = instance
    .W_commitments
    .iter()
    .map(|com| assign_and_absorb_point!(com))
    .collect::<Result<Vec<_>, _>>()?;
```

cyphersnake commented on src/ivc/fold_relaxed_plonk_instance_chip.rs:
```diff
@@ -779,7 +795,7 @@ mod tests {
         let chip = FoldRelaxedPlonkInstanceChip::<C1>::from_instance(
             PlonkInstance {
-                W_commitment: C1::random(&mut rnd),
+                W_commitments: vec![C1::random(&mut rnd)],
```
Why only one element?

chaosma replied:
this depends on how you test it. this is just quick hack as I mentioned before.

cyphersnake commented on src/ivc/fold_relaxed_plonk_instance_chip.rs:
```diff
@@ -732,7 +748,7 @@ where
             folded_X0: assigned_X0,
             folded_X1: assigned_X1,
-            input_W_commitment: assigned_instance_W_commitment_coordinates,
+            input_W_commitments: vec![assigned_instance_W_commitment_coordinates],
```
Look at comment above and use all `W` here

chaosma replied:
yeah, fixed it based on above comment

cyphersnake commented on src/plonk/mod.rs:
```diff
@@ -478,115 +554,203 @@ impl<F: PrimeField> TableData<F> {
-    pub fn plonk_instance<C: CurveAffine<ScalarExt = F>>(
+    /// run 0-round special soundness protocol
+    /// i.e no need to generate challenge
+    pub fn run_sps_protocol_0<C: CurveAffine<ScalarExt = F>>(
```
We should choose better names then `_protocol_N` & add context about cases of usage 
Maybe some enum with short names & descriptions, maybe just fn with long name, but just numbers confuse

chaosma replied:
I have updated the use cases in #34

chaosma replied:
I will resolve it for now. And let me know if you have a better naming.

cyphersnake replied:
We need a description of the pipeline depending on the scenario. 

I.e. we have quite clear situations and need to express them through code. I'll try to do that today

chaosma replied:
I have added the rest cases and one main function to call them in #73

cyphersnake commented on src/plonk/mod.rs:
```diff
@@ -478,115 +554,203 @@ impl<F: PrimeField> TableData<F> {
+            "should call TableData.assembly() first"
+        );
+
+        let W1 = concatenate_with_padding(&self.advice_columns[..], 2usize.pow(self.k));
```
You almost never need `[..]`

chaosma replied:
yeah, sometimes will forget, bad habit...

cyphersnake commented on src/util.rs:
```diff
@@ -175,6 +175,16 @@ pub(crate) fn normalize_trailing_zeros(bits: &mut Vec<bool>, bit_len: usize) {
+    let mut vs = vs.to_vec();
+    vs.iter_mut()
+        .flat_map(|v_i| {
+            v_i.resize(pad_size, F::ZERO);
+            v_i.drain(..)
```
Why is it so hard? What are you trying to accomplish?

cyphersnake replied:
I mean, there are a lot of downsides to using such a design, but I want to understand what effects you want to achieve in order to get the review as correct as possible

cyphersnake replied:
The simplest option with one memory allocation and no cloning at all is suggested, but maybe I missed something
```rust
pub(crate) fn concatenate_with_padding<F: PrimeField>(vs: &[Vec<F>], pad_size: usize) -> Vec<F> {
    vs.iter()
        .fold(Vec::with_capacity(vs.len() * pad_size), |mut result, v| {
            result.extend_from_slice(v);
            result.extend(iter::repeat(F::ZERO).take(pad_size.saturating_sub(v.len())));
            result
        })
}
```

chaosma replied:
No, there is no other concern. This is adapted from my previous code based on suggestion from chatgpt. The only purpose is to avoid extra clone.
What is the downside of using drain?

cyphersnake replied:
There was a lot of implicit clones there, as well as a lot of unnecessary allocations when expanding vectors.

cyphersnake replied:
@chaosma Do we need an error, if `vs` element is larger than padding size?

chaosma commented:
> I've finished the first round of reviewing and learning about SPS, but as always I will advocate the need for a general description in the code of some major points. Now, to understand what is going on in the code I need to read protostar paper in parallel, I would like to have at least some basic anchors and shortcuts.

I have updated the document here for reference: https://github.com/snarkify/sirius/issues/34

cyphersnake commented:
I have completely disassembled protostar and understand everything described here in general and in particular. Now I will try to parse the code with this knowledge and see if we can improve the accompanying part of this code.

cyphersnake commented on src/plonk/lookup.rs:
```diff
@@ -65,244 +52,155 @@ use crate::{
-    ) -> Vec<F> {
+    pub fn evaluate_m(&self, l: &[F], t: &[F]) -> Vec<F> {
         let mut m: Vec<F> = Vec::new();
         let mut processed_t = Vec::new();
-        for t_i in &ss.t {
+        for t_i in t {
             if processed_t.contains(&t_i) {
                 // If the current t_i has already been processed, push 0 to m
                 m.push(F::ZERO);
             } else {
-                let m_i = self.l.iter().filter(|l_j| *l_j == t_i).count() as u128;
+                let m_i = l.iter().filter(|l_j| *l_j == t_i).count() as u128;
                 m.push(F::from_u128(m_i));
                 processed_t.push(t_i);
             }
```
While I was understanding why there are no optimisations here (sorted `Vec` or `HashSet`), I refactored it a little bit

```rust
pub fn evaluate_m(&self, l: &[F], t: &[F]) -> Vec<F> {
    let mut processed_t = Vec::new();
                                                                                 
    t.iter()
        .map(|t_i| {
            if processed_t.contains(&t_i) {
                F::ZERO
            } else {
                processed_t.push(t_i);
                F::from_u128(l.iter().filter(|l_j| *l_j == t_i).count() as u128)
            }
        })
        .collect()
}
```

cyphersnake commented on src/plonk/lookup.rs:
```diff
@@ -313,52 +211,41 @@ impl<F: PrimeField> Witness<F> {
-                .collect::<Vec<F>>();
+        let g = t
+            .iter()
+            .zip(m)
```
`zip_eq`?

cyphersnake commented on src/plonk/lookup.rs:
```diff
@@ -313,52 +211,41 @@ impl<F: PrimeField> Witness<F> {
-    ) -> Result<(), String> {
-        let m = self.log_derivative_coeffs(ss);
-        let (h, g) = self.calc_inverse_terms(ss, r, &m);
+    pub fn is_sat(&self, l: &[F], t: &[F], r: F) -> Result<(), String> {
```
`is_satisfied`

chaosma replied:
I prefer the `is_sat` which is a common term in complexity theory. For example, 3-SAT is a term for boolean satisfiability.

cyphersnake commented on src/plonk/mod.rs:
```diff
@@ -49,31 +48,45 @@ pub struct PlonkStructure<C: CurveAffine> {
+    /// r2: challenge to calculate h and g in log-derivative relation
+    /// r3: combine all custom gates (P_i) and lookup relations (L_i), e.g.:
+    /// (P_1, P_2, L_1, L_2) -> P_1 + r3*P_2 + r3^2*L_1 + r3^3*L_2
     pub(crate) challenges: Vec<C::ScalarExt>,
```
We have many places with implicit size restrictions. I suggest for [discussion](https://github.com/snarkify/sirius/issues/72) to start using tinyvec crate, which allows to limit the maximum length. It would be hard enough to keep track of the potential invariants of such a large number of parameters and modify them, it would be better to do it with a compiler

chaosma replied:
sounds good. After I finish lookup, I can replace vec to tiny vector for optimization.

cyphersnake reviewed (changes requested):
I've generally refactored `run_sps_*`, but before I finish I need to understand why you don't have a check anywhere to see if a particular set of SPS protcol rounds are running in the correct invarainment?

I take it the work isn't finished yet? In #34 you described some scenarios, how will they be reflected in the code? 

Apart from the `run_sps_protocol_N` themselves everything else is trivial, but the design of these functions still confuses me. Like I don't see full picture inside codebase, like call of `*_3` fn

chaosma commented:
> I've generally refactored `run_sps_*`, but before I finish I need to understand why you don't have a check anywhere to see if a particular set of SPS protcol rounds are running in the correct invarainment?
> 
> I take it the work isn't finished yet? In #34 you described some scenarios, how will they be reflected in the code?
> 
> Apart from the `run_sps_protocol_N` themselves everything else is trivial, but the design of these functions still confuses me. Like I don't see full picture inside codebase, like call of `*_3` fn

It will be included in the next PR #73. My original plan is to continue working on #73. Since there are still several new comments to address. I will address them in the new PR. @cyphersnake
